spring-cloud-aws-s3-parent/spring-cloud-aws-s3/src/main/java/io/awspring/cloud/s3/InMemoryBufferingS3OutputStream.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * Copyright 2013-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.awspring.cloud.s3;
+
+import java.io.ByteArrayOutputStream;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Base64;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Optional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.lang.Nullable;
+
+import software.amazon.awssdk.core.exception.SdkException;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.services.s3.S3Client;
+import software.amazon.awssdk.services.s3.model.*;
+
+/**
+ * {@link S3OutputStream} implementation that buffers content to an internal {@link ByteArrayOutputStream} and streams
+ * the content as a MultiPartUpload as the buffer fills up. If the buffer's capacity is never exceeded, then an ordinary
+ * {@link PutObjectRequest} is used for efficiency. This implementation is thread-safe, but makes no guarantees to
+ * prevent data from interleaving. <br>
+ * <br>
+ * This stream <b>must</b> be closed to complete the upload.
+ *
+ * @author Sam Garfinkel
+ * @since 3.0
+ */
+public class InMemoryBufferingS3OutputStream extends S3OutputStream {
+
+	private static final Logger logger = LoggerFactory.getLogger(InMemoryBufferingS3OutputStream.class);
+
+	public static final int DEFAULT_BUFFER_CAPACITY = 1024 * 1024 * 5;
+
+	private final Location location;
+
+	private final S3Client s3Client;
+
+	@Nullable
+	private final ObjectMetadata objectMetadata;
+
+	@Nullable
+	private final S3ObjectContentTypeResolver contentTypeResolver;
+
+	private final int bufferSize;
```

Comment:
Perhaps we could either use `org.springframework.util.unit.DataSize` type or make it more explicit `bufferSizeInBytes`. I guess `DataSize` is a better option. Same applies to all the fields or parameters where we referred to bytes implicitly like `DEFAULT_BUFFER_CAPACITY`.